### PR TITLE
[TensorDec] update mode with bounding boxes

### PR DESCRIPTION
**Pull request by jinhyuck-park**
change mode type from const char to uint
and add bounding boxes mode

Signed-off-by: jinhyuck-park <jinhyuck83.park@samsung.com>

**Self evaluation:**
1. Build test: [*]Passed [ ]Failed [ ]Skipped
2. Run test: [*]Passed [ ]Failed [ ]Skipped
push origin <your_branch_name>
```

### CODE REVIEW

File: gst/tensor_decoder/tensordec.c [hunk 1]
```diff
@@ -512,10 +521,16 @@ gst_tensordec_set_property (GObject * object, guint prop_id,
...
       break;
     case PROP_MODE:
-      self->mode = g_value_dup_string (value);
+      if (g_strcmp0 (g_value_dup_string (value), "image_labeling") == 0) {
+        self->mode = IMAGE_LABELING;
+      } else if (g_strcmp0 (g_value_dup_string (value), "boundig_boxes") == 0) {
+        self->mode = BOUNDING_BOXES;
+      } else {
+        self->mode = DIRECT_VIDEO;
+      }
```
Comment (myungjoo) [inline gst/tensor_decoder/tensordec.c]: Reuse the internal API, ```find_key_strv()``` in tensor_common.h.

With
```
static const gchar *mode_names[] = {
  "direct_video",
  "image_labeling",
  "bounding_boxes",
  NULL
};
```
, you can have ```self->mode = find_key_strv (mode_names, key_name);```

Besides, ```g_value_dup_string()``` does "malloc". Make sure that the returned value is deallocated.

File: gst/tensor_decoder/tensordec.c [hunk 2]
```diff
@@ -543,7 +558,7 @@ gst_tensordec_get_property (GObject * object, guint prop_id,
       g_value_set_boolean (value, self->silent);
       break;
     case PROP_MODE:
-      g_value_set_string (value, self->mode);
+      g_value_set_uint (value, self->mode);
```
Comment (myungjoo) [inline gst/tensor_decoder/tensordec.c]: It's supposed to be string.

Having uint/int value for ```self->mode``` is great (it's highly recommended to do so), however, the get_property needs to return the same type.

File: gst/tensor_decoder/tensordec.c [hunk 3]
```diff
@@ -813,7 +828,7 @@ gst_tensordec_transform (GstBaseTransform * trans,
       res = gst_tensordec_copy_buffer (self, inbuf, outbuf);
       break;
     case OUTPUT_TEXT:
-      if (g_strcmp0 (self->mode, "image_labeling") == 0) {
+      if (self->mode == IMAGE_LABELING) {
```
Comment (myungjoo) [inline gst/tensor_decoder/tensordec.c]: :+1:

File: gst/tensor_decoder/tensordec.h [hunk 4]
```diff
@@ -70,7 +70,7 @@ struct _GstTensorDec
   gboolean add_padding; /**< If TRUE, zero-padding must be added during transform */
   gboolean silent; /**< True if logging is minimized */
   guint output_type; /**< Denotes the output type */
-  const gchar *mode; /** Mode for tensor decoder "direct_video" or "image_labeling" */
+  guint mode; /** Mode for tensor decoder "direct_video" or "image_labeling" or "bounding_boxes */
```
Comment (myungjoo) [inline gst/tensor_decoder/tensordec.h]: :+1: This is a good change!

File: gst/tensor_decoder/tensordec.c (hunk 1 above)
Comment (jinhyuck-park) [reply gst/tensor_decoder/tensordec.c]: I updated code with your recommands
including g_value_dup_string() deallocation issue

Comment (myungjoo) [reply gst/tensor_decoder/tensordec.c]: You don't need these ```if-else``` statements.
Remove Line 527-533, replace it with ```self->mode = find_key_strv (mode_names, temp_string);```

File: gst/tensor_decoder/tensordec.c [hunk 5]
```diff
@@ -106,6 +106,16 @@ enum
...
 
+/**
+ * @brief Decoder Mode.
+ */
+enum
+{
+  DIRECT_VIDEO,
+  IMAGE_LABELING,
+  BOUNDING_BOXES
+};
```
Comment (myungjoo) [inline gst/tensor_decoder/tensordec.c]: For readability, move this to tensordec.h near the lines of ```static const gchar *mode_names[] = { ...```

File: gst/tensor_decoder/tensordec.c [hunk 6]
```diff
@@ -1004,7 +1013,7 @@ gst_tensordec_set_caps (GstBaseTransform * trans,
   silent_debug_caps (incaps, "from incaps");
   silent_debug_caps (outcaps, "from outcaps");
 
-  if (g_strcmp0 (self->mode, "image_labeling") == 0) {
+  if (self->mode == IMAGE_LABELING) {
```
Comment (jaeyun-jung) [inline gst/tensor_decoder/tensordec.c]: line 1016 ~ 1029 is unnecessary.

line 1031 ~ 1040, this lines are to configure tensor info actually. 
IMO, it is better to set in-place mode false in line 1042 with the condition.
if mode==labeling then in-place=false
else if mode==video then in-place=!self->add_padding
else in-place=true

File: gst/tensor_decoder/tensordec.c [hunk 7]
```diff
@@ -512,10 +513,12 @@ gst_tensordec_set_property (GObject * object, guint prop_id,
       self->silent = g_value_get_boolean (value);
       break;
     case PROP_MODE:
-      self->mode = g_value_dup_string (value);
+      temp_string = g_value_dup_string (value);
+      self->mode = find_key_strv (mode_names, temp_string);
```
[the remark this answers is not in the document]
Comment (jinhyuck-park) [reply gst/tensor_decoder/tensordec.c]: I added exeption case with your recommand

File: gst/tensor_decoder/tensordec.c [hunk 8]
```diff
@@ -543,7 +546,13 @@ gst_tensordec_get_property (GObject * object, guint prop_id,
       g_value_set_boolean (value, self->silent);
       break;
     case PROP_MODE:
-      g_value_set_string (value, self->mode);
+      if (self->mode == DIRECT_VIDEO) {
```
[the remark this answers is not in the document]
Comment (jinhyuck-park) [reply gst/tensor_decoder/tensordec.c]: added as your recommand

File: gst/tensor_decoder/tensordec.c (hunk 6 above)
Comment (jinhyuck-park) [reply gst/tensor_decoder/tensordec.c]: I refactored codes as your recommand

File: gst/tensor_decoder/tensordec.c [hunk 9]
```diff
@@ -1004,37 +1011,26 @@ gst_tensordec_set_caps (GstBaseTransform * trans,
...
-      }
+    if (gst_tensor_config_from_structure (&config, s) &&
+        gst_tensordec_check_consistency (self, &config)) {
+      self->negotiated = TRUE;
     }
-
+  }
+  if (self->mode == IMAGE_LABELING) {
+    gst_base_transform_set_in_place (trans, FALSE);
+    self->negotiated = TRUE;
```
Comment (jaeyun-jung) [inline gst/tensor_decoder/tensordec.c]: Please remove this line.
(this flag 'negotiated' is set in line 1021. if negotiated is false, plugin caps negotiation is failed.)

Comment (jinhyuck-park) [reply gst/tensor_decoder/tensordec.c]: for this problem,
I added 	(config->rate_n = 30; config->info.dimension[0] = 1001;)
without this, text caps negotiation fails

Comment (jinhyuck-park) [reply gst/tensor_decoder/tensordec.c]: for this problem,
I added 	(config->rate_n = 30; config->info.dimension[0] = 1001;)
without this, text caps negotiation fails

Comment (jinhyuck-park) [reply gst/tensor_decoder/tensordec.c]: for this problem,
I added 	(config->rate_n = 30; config->info.dimension[0] = 1001;)
without this, text caps negotiation fails

**Pull request merged**